internal/expr/bindtypes.go

```diff
@@ -44,80 +41,113 @@ func (pe *ParsedExpr) BindTypes(args ...any) (te *TypedExpr, err error) {
 		return nil, err
 	}
 
-	var sql bytes.Buffer
-
-	var inCount int
-	var outCount int
-
-	var outputs = make([]typeinfo.Output, 0)
-	var inputs = make([]typeinfo.Input, 0)
-
-	var outputUsed = make(map[typeinfo.Output]bool)
+	// Bind types to each expression.
+	typedExprs := []typedExpression{}
+	outputUsed := map[typeinfo.Output]bool{}
+	for _, expr := range *pe {
+		te, err := expr.bindTypes(argInfo)
+		if err != nil {
+			return nil, err
+		}
 
-	// Check and expand each query expr.
-	for _, expr := range pe.exprs {
-		switch e := expr.(type) {
-		case *inputExpr:
-			inLoc, err := bindInputTypes(argInfo, e)
-			if err != nil {
-				return nil, err
-			}
-			sql.WriteString("@sqlair_" + strconv.Itoa(inCount))
-			inCount++
-			inputs = append(inputs, inLoc)
-		case *outputExpr:
-			outCols, os, err := bindOutputTypes(argInfo, e)
+		if toe, ok := te.(*typedOutputExpr); ok {
+			err = trackUsedOutputs(outputUsed, toe)
 			if err != nil {
 				return nil, err
 			}
+		}
 
-			for _, o := range os {
-				if ok := outputUsed[o]; ok {
-					return nil, fmt.Errorf("%s appears more than once in output expressions", o.String())
-				}
-				outputUsed[o] = true
-			}
+		typedExprs = append(typedExprs, te)
+	}
 
-			for i, c := range outCols {
-				sql.WriteString(c.String())
-				sql.WriteString(" AS ")
-				sql.WriteString(markerName(outCount))
-				if i != len(outCols)-1 {
-					sql.WriteString(", ")
-				}
-				outCount++
-			}
-			outputs = append(outputs, os...)
-		case *bypass:
-			sql.WriteString(e.chunk)
-		default:
-			return nil, fmt.Errorf("internal error: unknown query expr type %T", expr)
+	typedExpr := TypeBoundExpr(typedExprs)
+	return &typedExpr, nil
+}
+
+// trackUsedOutputs tracks the outputs used already and returns an error if an
+// output has been used twice. It updates outputUsed with the outputs from the
+// typedOutputExpr.
+func trackUsedOutputs(outputUsed map[typeinfo.Output]bool, toe *typedOutputExpr) error {
```

Comment:
I am not sure this should be a separate function, to me it is clearer reading the 6 lines than having to click on the IDE to see the function comment and read it.

Reply:
I've inlined it again. I'd originally made it a method of the `typedOutputExpr` because I didn't want to break abstraction levels in the `BindInputs` function (since this relies on the internals of the `typedOutputExpr`), but I agree that as a standalone function it doesn't do much.